File: app/src/main/java/com/muparse/PlaylistAdapter.java
package com.muparse;

/*
  Created by fedor on 28.11.2016.
 */


import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.amulyakhare.textdrawable.TextDrawable;
import com.amulyakhare.textdrawable.util.ColorGenerator;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;

public class PlaylistAdapter extends RecyclerView.Adapter<PlaylistAdapter.ItemHolder> implements Filterable {

    private final Context mContext;
    private final LayoutInflater mInflater;
    private List<M3UItem> mItem = new ArrayList<>();
    private TextDrawable textDrawable;
    private ColorGenerator generator = ColorGenerator.MATERIAL;

    public PlaylistAdapter(Context c) {
        mContext = c;
        mInflater = LayoutInflater.from(mContext);
    }

    @Override
    public ItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View sView = mInflater.inflate(R.layout.item_playlist, parent, false);
        return new ItemHolder(sView);
    }

    @Override
    public int getItemViewType(int position) {
        return 0;
    }

    @Override
    public void onBindViewHolder(final ItemHolder holder, final int position) {
        final M3UItem item = mItem.get(position);
        if (item != null) {
            holder.update(item);
        }
    }

    @Override
    public int getItemCount() {
        return mItem.size();
    }

    void update(List<M3UItem> _list) {
        this.mItem = _list;
        notifyDataSetChanged();
    }

    @Override
    public Filter getFilter() {
        return new Filter() { //TODO search it on github
            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                mItem.clear();
                //mItem.addAll((ArrayList<M3UItem>) results.values);
                notifyDataSetChanged();
            }
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                FilterResults results = new FilterResults();
                if (!(constraint.length() == 0)){
                    mItem.clear();
                    final String filtePatt = constraint.toString().toLowerCase().trim();
                    for (M3UItem itm: mItem){
                        if (itm.getItemName().toLowerCase().contains(filtePatt)){
                            Toast.makeText(mContext, "Google", Toast.LENGTH_SHORT).show();
                            mItem.add(itm);
                        }
                        mItem.add(itm);
                    }
                }
                results.values = mItem;
                results.count = mItem.size();
                return results;
            }
        };
    }

    public class ItemHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        final PackageManager pm = mContext.getPackageManager();
        final boolean isApp = Utils.getInstance().isPackageInstalled(pm);
        TextView name;
        ImageView cImg;

        ItemHolder(View view) {
            super(view);
            view.setOnClickListener(this);
            name = view.findViewById(R.id.item_name);
            cImg = view.findViewById(R.id.cimg);
        }

        void update(final M3UItem item) {
            try {
            name.setText(item.getItemName());
                int color = generator.getRandomColor();
                if (item.getItemIcon().isEmpty()) {
                    textDrawable = TextDrawable.builder()
                            .buildRoundRect(String.valueOf(item.getItemName().charAt(0)), color, 100);
                    cImg.setImageDrawable(textDrawable);
                } else {
                    if (Utils.getInstance().isNetworkAvailable(mContext)) {
                        Picasso.with(mContext).load(item.getItemIcon()).into(cImg);
                    } else {
                        textDrawable = TextDrawable.builder()
                                .buildRoundRect(String.valueOf(item.getItemName().charAt(0)), color, 100);
                        cImg.setImageDrawable(textDrawable);
                    }
                }
            } catch (Exception ignored) {
            }
        }

        public void onClick(View v) {
            try {
                int position = getLayoutPosition();
                final M3UItem imm = mItem.get(position);
//                if (isApp) { //isApplicationInstalled
//                    playy(imm.getItemUrl(), imm.getItemName());
//                } else { // isApplicationNotFound
//                    playerNotFound(mContext,"iPtv","Player not found. Install MX Player.");
//                }
                Intent intent = new Intent(mContext, playerExo.class);
                intent.putExtra("Name", imm.getItemName());
                intent.putExtra("Url", imm.getItemUrl());
                Log.e("Google", imm.getItemUrl());
                mContext.startActivity(intent);
            } catch (Exception ignored) {}
        }

        void playy(String urli, String channel) {
            try {
                Uri videoUri = Uri.parse(urli);
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setDataAndType(videoUri, "application/x-mpegURL");
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                intent.setPackage("com.mxtech.videoplayer.ad");
                intent.putExtra("title", channel);
                mContext.startActivity(intent);
            } catch (Exception ignored) {}
        }

        void playerNotFound(Context context,String title,String message) {
            AlertDialog.Builder localBuilder = new AlertDialog.Builder(context);
            localBuilder.setTitle(title);
            localBuilder.setMessage(message);
            localBuilder.setPositiveButton("Install", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface paramAnonymousDialogInterface, int paramAnonymousInt) {
                    try {
                        mContext.startActivity(new Intent("android.intent.action.VIEW", Uri.parse("market://details?id=com.mxtech.videoplayer.ad")));
                    } catch (ActivityNotFoundException ActivityNotFoundException) {
                        mContext.startActivity(new Intent("android.intent.action.VIEW", Uri.parse("https://play.google.com/store/apps/details?id=com.mxtech.videoplayer.ad&hl=en")));
                    }
                }
            });
//        localBuilder.setNegativeButton("Google", new DialogInterface.OnClickListener()
//        {
//            public void onClick(DialogInterface paramAnonymousDialogInterface, int paramAnonymousInt) {}
//        });
            //localBuilder.setCancelable(false);
            localBuilder.create().show();
        }
    }
}

File: app/src/main/java/com/muparse/PlaylistAdapter.kt
package com.muparse

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.RecyclerView
import com.amulyakhare.textdrawable.TextDrawable
import com.amulyakhare.textdrawable.util.ColorGenerator
import com.bumptech.glide.Glide
import java.util.*

class PlaylistAdapter internal constructor(private val mContext: Context) : RecyclerView.Adapter<PlaylistAdapter.ItemHolder>(), Filterable {
    private lateinit var mItem: MutableList<M3UItem>
    private var filteredList: MutableList<M3UItem> = ArrayList()
    private var textDrawable: TextDrawable? = null
    private val generator = ColorGenerator.MATERIAL
    private val playerId = 0
    private val pm = mContext.packageManager

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemHolder {
        val sView = LayoutInflater.from(parent.context).inflate(R.layout.item_playlist, parent, false)
        return ItemHolder(sView)
    }

    override fun getItemViewType(position: Int): Int {
        return 0
    }

    override fun onBindViewHolder(holder: ItemHolder, position: Int) {
        mItem[position].apply {
            holder.update(this)

            holder.rel.setOnClickListener {
                when (playerId) {
                    0 -> {
                        val intent = Intent(mContext, PlayerExo::class.java)
                        intent.putExtra("Name", itemName)
                        intent.putExtra("Url", itemUrl)
                        Log.e("Google", itemUrl)
                        mContext.startActivity(intent)
                    }
                    1 -> {
                        val isApp = Utils.getInstance().isPackageInstalled(pm)
                        if (isApp) {
                            holder.playy(itemUrl!!, itemName!!)
                        } else { // isApplicationNotFound
                            holder.playerNotFound(mContext, "iPtv", "Player not found. Install MX Player.")
                        }
                    }
                }
            }

        }
    }

    override fun getItemCount(): Int {
        return mItem.size
    }

    fun update(_list: MutableList<M3UItem>) {
        this.mItem = _list
        notifyDataSetChanged()
    }

    override fun getFilter(): Filter {
        return object : Filter() { //TODO search it on github
            override fun publishResults(constraint: CharSequence, results: FilterResults) {
//                mItem.clear()
                //mItem.addAll((ArrayList<M3UItem>) results.values);
                notifyDataSetChanged()
            }

            override fun performFiltering(constraint: CharSequence): FilterResults {
                filteredList = mItem
                return FilterResults().apply {
                    if (constraint.isNotEmpty()) {
                        filteredList.clear()
                        val filtePatt = constraint.toString().toLowerCase().trim { it <= ' ' }
                        for (itm in mItem) {
                            if (itm.itemName!!.toLowerCase().contains(filtePatt)) {
                                Toast.makeText(mContext, "Google", Toast.LENGTH_SHORT).show()
                                filteredList.add(itm)
                            }
//                            mItem.add(itm)
                        }
                    }
                    values = filteredList
                    count = filteredList.size
                }

            }
        }
    }

    inner class ItemHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {

        internal var name: TextView = view.findViewById(R.id.item_name)
        private var cImg: ImageView = view.findViewById(R.id.cimg)
        var rel: RelativeLayout = view.findViewById(R.id.relLayout)

        internal fun update(item: M3UItem) {
            try {
                name.text = item.itemName
                val color = generator.randomColor
                if (item.itemIcon!!.isEmpty()) {
                    textDrawable = TextDrawable.builder()
                            .buildRoundRect(item.itemName!![0].toString(), color, 100)
                    cImg.setImageDrawable(textDrawable)
                } else {
                    if (Utils.getInstance().isNetworkAvailable(mContext)) {
                        Glide.with(mContext).load(item.itemIcon).into(cImg)
                    } else {
                        textDrawable = TextDrawable.builder()
                                .buildRoundRect(item.itemName!![0].toString(), color, 100)
                        cImg.setImageDrawable(textDrawable)
                    }
                }
            } catch (ignored: Exception) {
            }

        }

        internal fun playy(urli: String, channel: String) {
            try {
                val videoUri = Uri.parse(urli)
                val intent = Intent(Intent.ACTION_VIEW)
                intent.setDataAndType(videoUri, "application/x-mpegURL")
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                intent.setPackage("com.mxtech.videoplayer.ad")
                intent.putExtra("title", channel)
                mContext.startActivity(intent)
            } catch (ignored: Exception) {
            }

        }

        internal fun playerNotFound(context: Context, title: String, message: String) {
            val localBuilder = AlertDialog.Builder(context)
            localBuilder.setTitle(title)
            localBuilder.setMessage(message)
            localBuilder.setPositiveButton("Install") { paramAnonymousDialogInterface, paramAnonymousInt ->
                try {
                    mContext.startActivity(Intent("android.intent.action.VIEW", Uri.parse("market://details?id=com.mxtech.videoplayer.ad")))
                } catch (ActivityNotFoundException: ActivityNotFoundException) {
                    mContext.startActivity(Intent("android.intent.action.VIEW", Uri.parse("https://play.google.com/store/apps/details?id=com.mxtech.videoplayer.ad&hl=en")))
                }
            }
            //        localBuilder.setNegativeButton("Google", new DialogInterface.OnClickListener()
            //        {
            //            public void onClick(DialogInterface paramAnonymousDialogInterface, int paramAnonymousInt) {}
            //        });
            //localBuilder.setCancelable(false);
            localBuilder.create().show()
        }
    }
}
